[OpInfo] Make SkipInfo assert failure by default

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62725

This PR updates SkipInfo to assert that the test being skipped raises an Error by default. This can be disabled with `expected_failure=False`. This has the following benefits:
- Test skips have to target the exact tests that are failing (e.g. only failing on CUDA)
- When things get fixed, these tests will fail so it reminds us to remove the skip
- We can distinguish between tests that are expected to fail from tests that are being skipped because they are slow or for some other reason.

### Note
skips for test_reference_numerics tests for UnaryUFuncInfo have expected_failure=False for now because many of these should be changed to a toleranceOverride decorator.

Differential Revision: [D30458677](https://our.internmc.facebook.com/intern/diff/D30458677)

heitorschueroff committed:
[OpInfo] Make SkipInfo assert failure by default

[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -7440,9 +7397,9 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
-                       SkipInfo('TestGradients'),
-                       SkipInfo('TestJit'),
-                       SkipInfo('TestCommon'),
+                       SkipInfo('TestGradients', expected_failure=False),
+                       SkipInfo('TestJit', expected_failure=False),
+                       SkipInfo('TestCommon', expected_failure=False),
```
If you have a hard time landing this due to logical merge conflicts, then it might be better to default to expect_failure=False and then slowly flip the switch

heitorschueroff replied:
Good suggestion. I'll split this into a PR adding the kwarg and another setting it to True

zou3519 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -70,15 +70,24 @@ def is_active(self, cls_name, test_name, device_type, dtype):
-    """Describes which test, or type of tests, should be skipped when testing
-       an operator. Any test that matches all provided arguments will be skipped.
-       The skip will only be checked if the active_if argument is True."""
+    """Information about which tests to skip."""
```
I would keep the first sentence of the blurb to make it clearer that SkipInfo is used for operator testing.

zou3519 reviewed (approved):
This looks good to me, modulo the test failures. I added a suggestion for how to land this without worrying about people merge-conflicting with you.

Are there any cases where tests are skipped because they fail non-deterministically?

heitorschueroff committed:
Update on "[OpInfo] Make SkipInfo assert failure by default"


This PR updates SkipInfo to assert that the test being skipped raises an Error by default. This can be disabled with `expected_failure=False`. This has the following benefits:
- Test skips have to target the exact tests that are failing (e.g. only failing on CUDA)
- When things get fixed, these tests will fail so it reminds us to remove the skip
- We can distinguish between tests that are expected to fail from tests that are being skipped because they are slow or for some other reason.

[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Make SkipInfo assert failure by default"


This PR updates SkipInfo to assert that the test being skipped raises an Error by default. This can be disabled with `expected_failure=False`. This has the following benefits:
- Test skips have to target the exact tests that are failing (e.g. only failing on CUDA)
- When things get fixed, these tests will fail so it reminds us to remove the skip
- We can distinguish between tests that are expected to fail from tests that are being skipped because they are slow or for some other reason.

[ghstack-poisoned]

mruberry commented:
Looks like there are still some tests to review here

heitorschueroff committed:
Update on "[OpInfo] Make SkipInfo assert failure by default"


This PR updates SkipInfo to assert that the test being skipped raises an Error by default. This can be disabled with `expected_failure=False`. This has the following benefits:
- Test skips have to target the exact tests that are failing (e.g. only failing on CUDA)
- When things get fixed, these tests will fail so it reminds us to remove the skip
- We can distinguish between tests that are expected to fail from tests that are being skipped because they are slow or for some other reason.

[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Make SkipInfo assert failure by default"


This PR updates SkipInfo to assert that the test being skipped raises an Error by default. This can be disabled with `expected_failure=False`. This has the following benefits:
- Test skips have to target the exact tests that are failing (e.g. only failing on CUDA)
- When things get fixed, these tests will fail so it reminds us to remove the skip
- We can distinguish between tests that are expected to fail from tests that are being skipped because they are slow or for some other reason.

[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Make SkipInfo assert failure by default"


This PR updates SkipInfo to assert that the test being skipped raises an Error by default. This can be disabled with `expected_failure=False`. This has the following benefits:
- Test skips have to target the exact tests that are failing (e.g. only failing on CUDA)
- When things get fixed, these tests will fail so it reminds us to remove the skip
- We can distinguish between tests that are expected to fail from tests that are being skipped because they are slow or for some other reason.

[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Make SkipInfo assert failure by default"


This PR updates SkipInfo to assert that the test being skipped raises an Error by default. This can be disabled with `expected_failure=False`. This has the following benefits:
- Test skips have to target the exact tests that are failing (e.g. only failing on CUDA)
- When things get fixed, these tests will fail so it reminds us to remove the skip
- We can distinguish between tests that are expected to fail from tests that are being skipped because they are slow or for some other reason.

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30458677).

heitorschueroff committed:
Update on "[OpInfo] Make SkipInfo assert failure by default"


This PR updates SkipInfo to assert that the test being skipped raises an Error by default. This can be disabled with `expected_failure=False`. This has the following benefits:
- Test skips have to target the exact tests that are failing (e.g. only failing on CUDA)
- When things get fixed, these tests will fail so it reminds us to remove the skip
- We can distinguish between tests that are expected to fail from tests that are being skipped because they are slow or for some other reason.

Differential Revision: [D30458677](https://our.internmc.facebook.com/intern/diff/D30458677)

[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Make SkipInfo assert failure by default"


This PR updates SkipInfo to assert that the test being skipped raises an Error by default. This can be disabled with `expected_failure=False`. This has the following benefits:
- Test skips have to target the exact tests that are failing (e.g. only failing on CUDA)
- When things get fixed, these tests will fail so it reminds us to remove the skip
- We can distinguish between tests that are expected to fail from tests that are being skipped because they are slow or for some other reason.

Differential Revision: [D30458677](https://our.internmc.facebook.com/intern/diff/D30458677)

[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Make SkipInfo assert failure by default"


This PR updates SkipInfo to assert that the test being skipped raises an Error by default. This can be disabled with `expected_failure=False`. This has the following benefits:
- Test skips have to target the exact tests that are failing (e.g. only failing on CUDA)
- When things get fixed, these tests will fail so it reminds us to remove the skip
- We can distinguish between tests that are expected to fail from tests that are being skipped because they are slow or for some other reason.

Differential Revision: [D30458677](https://our.internmc.facebook.com/intern/diff/D30458677)

[ghstack-poisoned]

ezyang commented:
@mruberry mentioned this in our QS sync, and I wanted to drive by and say I don't understand the motivation for this patch. Forgetting about the OpInfo specific API for a moment, if you want to assert that something is a failure, you should use `expectedFailure`, not `skip`. Running a test *anyway* even if someone told you it should be skipped is a great way to, e.g., (1) hard crash the test suite, because maybe you skipped the test because it was unsafe to run, or (2) add a bunch of flaky errors, because the test is skipped because sometimes it passes and sometimes it doesn't.

What would make more sense to me, perhaps, is a version of skip that runs the test anyway but ignores the result, and then some reporting mechanism on master that can let us know when a test is being run, its result is being ignored, but the result is actually always success or failure (and then we can go in and make it more precise in that situation).

zou3519 commented:
> Forgetting about the OpInfo specific API for a moment, if you want to assert that something is a failure, you should use expectedFailure, not skip

Most of the things that are being skipped today (the deterministic and non-crashy test failures) are cases where the developer should have used an expectedFailure but didn't because there wasn't an easy way to do it so they chose to use SkipInfo.

> What would make more sense to me, perhaps, is a version of skip that runs the test anyway but ignores the result, and then some reporting mechanism on master that can let us know when a test is being run, its result is being ignored, but the result is actually always success or failure (and then we can go in and make it more precise in that situation).

Perhaps what we want here is separate SkipInfo and ExpectedFailureInfo decorators to make the semantics clearer and let the author of the OpInfo (or related tests) decide what semantics they want.

If we separate our SkipInfo into SkipInfo and ExpectedFailureInfo then devs can clearly mark the cases where they skipped due to crashes or non-determinism and other places as expected failures

mruberry commented:
More stuff always complicates things and attention is hyper limited and OpInfos are often perceived as too complex already, but sure we can extend `skips` to take both SkipInfos and `ExpectedFailureInfos` (we might want to find another name). Technically we could just make `ExpectedFailureInfo` a subclass of SkipInfo that sets the expected failure kwarg to `True.`

ezyang commented:
> More stuff always complicates things and attention is hyper limited and OpInfos are often perceived as too complex already, but sure we can extend skips to take both SkipInfos and ExpectedFailureInfos (we might want to find another name).

A way to limit cognitive load is to make sure APIs follow prevailing ecosystem conventions. Python unittest module natively comes with "skip" and "expectedFailure" and so I would expect an average Python user to know these idioms. Of course, it would have been better if these decorators could have been used directly...

mruberry commented:
> > More stuff always complicates things and attention is hyper limited and OpInfos are often perceived as too complex already, but sure we can extend skips to take both SkipInfos and ExpectedFailureInfos (we might want to find another name).
> 
> A way to limit cognitive load is to make sure APIs follow prevailing ecosystem conventions. Python unittest module natively comes with "skip" and "expectedFailure" and so I would expect an average Python user to know these idioms. Of course, it would have been better if these decorators could have been used directly...

You actually can use the decorators directly but then they'll apply to every generated test, which is (generally) not we want for skipping/xfailing.

heitorschueroff commented:
If we are going to introduce a new `ExpectedFailureInfo` then should we also add a new `expected_failures` kwarg to `OpInfo`? Alternatively, we can leave the default of expected_failure kwarg in `SkipInfo` to `False` and change only where we want to assert failure.

mruberry commented:
> If we are going to introduce a new `ExpectedFailureInfo` then should we also add a new `expected_failures` kwarg to `OpInfo`? Alternatively, we can leave the default of expected_failure kwarg in `SkipInfo` to `False` and change only where we want to assert failure.

Instead of introducing new metadata let's reuse the existing "decorators" or "skips" fields. SkipInfos are actually just DecorateInfos, anyway, and we can model ExpectedFailureInfos the same way. So let's just use "ExpectedFailureInfo" as sugar for a DecorateInfo with an efail decorator. I would still start putting them under "skips" still, and not "decorators", to be clear that these are tests which are excluded for readability. This makes the name "skips" a bit of a misnomer and we could think about updating the name of that field to be "skips_and_expected_failures" but that's a lot of text.

ngimel commented:
Let's separate UX and functionality concern. Functionally, you want to sometimes skip a test, and sometimes run a test and make sure it fails. Typically, that is achieved by `@skip` and `@expectedFail` pytest decorators. #63481 adds this functionality to SkipInfo, and this PR makes the actual tests use it, expecting failure by default, or, for flaky/crashing tests, not expecting failure and instead skipping the test altogether. This is the functionality we agree we want and need. 
Now, let's discuss UX. Currently, OpInfo has an attribute `skips` and an attribute `decorators` that both can be used to skip the tests. There's DecorateInfo that is typically used in the `decorators` attribute that allows users to skip only some of the generated tests instead of all of them. There's SkipInfo, derived from DecorateInfo, whos only difference is that decorator is hardcoded to be either `expectedFailure` or `skip` (it used to be `skip` only until recently). The proposal above is to limit SkipInfo to only `skip`, and create `ExpectedFailureInfo` with a decorator hardcoded  to be `expectedFailure`. 
Personally, I dislike having an assortment of custom classes with the only distinguishing feature being a hardcoded attribute. It saves very little typing, and it increases cognitive load a lot. I also dislike `skips` and `decorators` attributes, for the same reason - one is a strict subset of functionality of another (see e.g. how `decorators` attribute is used to actually skip tests, #64469), that saves very little typing and doesn't make things any clearer. So why don't we leave `decorators` and `DecorateInfo` only, then developers can very explicitly express their intent on whether they want to skip or expect failure on the test?

heitorschueroff commented:
Closing this PR since we have removed SkipInfo.
